[Mvc] Support IAsyncDisposable for controllers and pages

javiercn opened the pull request:
[Fixes #13150]

 *Adds async versions of Release to the relevant interfaces in different
  factories and activators.
* The new methods offer a default interface implementation that will
  delegate to the syncrhonous version and just call dispose.
* Our implementations will override the default implementation and handle
  disposing IAsyncDisposable implementations.
* Our implementations will favor IAsyncDisposable over IDisposable when
  both interfaces are implemented.
* Extenders will have to override the new methods included to support
  IAsyncDisposable instances.

The general pattern is that there is a new method on the relevant interfaces that offers a default implementation that delegates to the sync version in the interface. The idea is that extenders will have to do work to support IAsyncDisposable too, but this way we avoid having a breaking change and libraries can add support at their own pace by overriding the default implementation.

We should consider marking the sync versions on the interface as `[Obsolete]` and point out to the new async versions.

Alternatively we could consider doing a breaking change here since it's very early in the release and we have time to advertise it.

The goal is to minimize the problems due to incompatible libraries for people updating to 6.0.

pranavkm reviewed (commented):
This looks greats. Are you planning to update this for view components:

* https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.ViewFeatures/src/ViewComponents/IViewComponentFactory.cs
* https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.ViewFeatures/src/ViewComponents/IViewComponentActivator.cs

It might also be helpful to throw if an IView instance implements IAsyncDisposable rather than silently no-op. It's invoked in sync code-paths and supporting IAsyncDisposable is far too hard there. (e.g. https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.ViewFeatures/src/TemplateRenderer.cs#L138)

javiercn commented:
@pranavkm those are great points. 

I was mainly testing the waters to root out any big blockers. The part that I'm more concerned about is the extensibility, but I think the approach I followed offers a reasonable trade-off. 

I think we should consider marking the existing sync releaser members as obsolete to hint authors to implement the async capable versions.

javiercn commented on src/Mvc/Mvc.Core/src/Infrastructure/ResourceInvoker.cs:
```diff
@@ -83,67 +83,24 @@ public virtual Task InvokeAsync()
-            {
-                return Task.FromException(scopeException!);
-            }
+            return ReleaseResourcesCore(scope);
```
I refactored this code because it wasn't consistent depending on whether the path was synchronous or async or whether the values were being logged.

pranavkm commented on src/Mvc/Mvc.Core/src/Infrastructure/ResourceInvoker.cs:
```diff
@@ -187,11 +143,77 @@ static async Task Logged(ResourceInvoker invoker)
             }
         }
 
+        internal Task ReleaseResourcesCore(IDisposable? scope)
```
`ValueTask`?

javiercn replied:
done

pranavkm commented on src/Mvc/Mvc.Core/src/Infrastructure/ResourceInvoker.cs:
```diff
@@ -187,11 +143,77 @@ static async Task Logged(ResourceInvoker invoker)
+                releaseException = exception;
+            }
+
+            if (releaseException == null && !releaseResult.IsCompletedSuccessfully)
```
Could we move this inside the `try` block? 

```C#
try
{
   var releaseResult = ReleaseResources();
   if (!releaseResult.IsCompletedSuccessfully)
   {
      return ReleaseResourcesAwaited(releaseResult);
   }
}
catch
```

javiercn replied:
done

pranavkm commented on src/Mvc/Mvc.Razor/src/RazorViewEngine.cs:
```diff
@@ -484,6 +484,11 @@ private ViewEngineResult CreateViewEngineResult(ViewLocationCacheResult result,
             var view = new RazorView(this, _pageActivator, viewStarts, page, _htmlEncoder, _diagnosticListener);
+            if (view is IAsyncDisposable)
+            {
+                throw new InvalidOperationException("Async disposable views are not supported.");
```
Resx file

javiercn replied:
done

pranavkm commented on src/Mvc/Mvc.RazorPages/test/Infrastructure/DefaultPageFactoryProviderTest.cs:
```diff
@@ -317,6 +393,26 @@ private static IPageActivatorProvider CreateActivator()
+                    };
+                });
+
+            ValueTask SyncDispose(IDisposable disposable) { disposable.Dispose(); return default; }
```
Formatting (separate lines).

javiercn replied:
done

javiercn committed:
Addressed feedback

javiercn committed:
[Mvc] Support IAsyncDisposable in controllers, pages and view components

* Adds async versions of Release to the relevant interfaces in different
  factories and activators.
* The new methods offer a default interface implementation that will
  delegate to the syncrhonous version and just call dispose.
* Our implementations will override the default implementation and handle
  disposing IAsyncDisposable implementations.
* Our implementations will favor IAsyncDisposable over IDisposable when
  both interfaces are implemented.
* Extenders will have to override the new methods included to support
  IAsyncDisposable instances.

pranavkm commented:
new API :|

javiercn commented:
@pranavkm It's life 😃

pranavkm commented on src/Mvc/Mvc.Core/src/Controllers/IControllerActivator.cs:
```diff
@@ -20,5 +22,16 @@ public interface IControllerActivator
         /// <param name="context">The <see cref="ControllerContext"/> for the executing action.</param>
         /// <param name="controller">The controller to release.</param>
         void Release(ControllerContext context, object controller);
```
Obsolete and never delete.
